api/observability/tracing/ssh/ssh.go

```diff
@@ -153,14 +154,85 @@ func NewClientConn(ctx context.Context, conn net.Conn, addr string, config *ssh.
 	)
 	defer span.End()
 
+	// ssh.ClientConfig.Timeout is not the total timeout for the connection
+	// establishment, including DNS resolution, TCP connection, but it doesn't
+	// include the SSH estalbishment.
+	// From the crypto/ssh docs:
+	// > Timeout is the maximum amount of time for the TCP connection to establish.
+	//
+	// Since we pass the connection here, the timeout will never be enforced by the
+	// ssh package. `NewClientConnWithDeadline` tries to enforced it by setting
+	// the read deadline on the connection, but might not be sufficient because
+	// we have some net.Conn implementations that don't support setting read deadlines
+	// and will block forever.
+	// To be sure that we don't block forever, we set up a timer that will close
+	// the connection when the timeout is reached.
+	// If the context has a deadline, we use that instead and take the minimum
+	// between the two.
+	// If neither is set, we default to 30 seconds.
+
+	// We aim to close the connection to avoid clients to hang forever
+	// if the server is not responding.
+	ctx, cancel := context.WithTimeout(
+		ctx,
+		getTimeoutOrDefault(ctx, config),
+	)
```

Comment:
There's no reason to go look at the parent context's deadline here, ctx will get canceled when that deadline hits anyway.

```suggestion
	if config.Timeout >= 0 {
		timeoutCtx, cancel := context.WithTimeout(ctx, cmp.Or(config.Timeout, defaults.DefaultIOTimeout)
		defer cancel()
		ctx = timeoutCtx
	}
```